bluemira/equilibria/opt_constraint_funcs.py

```diff
@@ -0,0 +1,315 @@
+    """
+    constraint[:] = objective_function(vector, grad) - maximum_fom
+    return constraint
+
+
+def Ax_b_constraint(constraint, vector, grad, a_mat, b_vec, value, scale):  # noqa: N802
+    """
+    Constraint function of the form:
+        A.x - b < value
+
+    Parameters
+    ----------
+    constraint: np.ndarray
+        Constraint array (modified in place)
+    vector: np.ndarray
+        Variable vector
+    grad: np.ndarray
+        Constraint Jacobian (modified in place)
+    a_mat: np.ndarray
+        Response matrix
+    b_vec: np.ndarray
+        Target value vector
+    value: float
+        Target constraint value
+    scale: float
+        Current scale with which to calculate the constraints
+    """
+    constraint[:] = np.dot(a_mat, scale * vector) - b_vec - value
+    if grad.size > 0:
+        grad[:] = scale * a_mat
+    return constraint
+
+
+def L2_norm_constraint(  # noqa: N802
+    constraint, vector, grad, a_mat, b_vec, value, scale
+):
+    """
+    Constrain the L2 norm of an Ax-b system of equations.
+    ||(Ax - b)||² < value
+
+    Parameters
+    ----------
+    constraint: np.ndarray
+        Constraint array (modified in place)
+    vector: np.ndarray
+        Variable vector
+    grad: np.ndarray
+        Constraint Jacobian (modified in place)
+    A_mat: np.ndarray
+        Response matrix
+    b_vec: np.ndarray
+        Target value vector
+    scale: float
+        Current scale with which to calculate the constraints
+
+    Returns
+    -------
+    constraint: np.ndarray
+        Updated constraint vector
+    """
+    vector = scale * vector
+    residual = a_mat @ vector - b_vec
+    constraint[:] = residual.T @ residual - value
+
+    if grad.size > 0:
+        grad[:] = 2 * scale * (a_mat.T @ a_mat @ vector - a_mat.T @ b_vec)
+
+    return constraint
+
+
+def current_midplane_constraint(
+    constraint, vector, grad, eq, radius, scale, inboard=True
+):
+    """
+    Constraint function to constrain the inboard or outboard midplane
+    of the plasma during optimisation.
+
+    Parameters
+    ----------
+    constraint: np.ndarray
+        Constraint array (modified in place)
+    vector: np.ndarray
+        Current vector
+    grad: np.ndarray
+        Constraint Jacobian (modified in place)
+    eq: Equilibrium
+        Equilibrium to use to fetch last closed flux surface from.
+    radius: float
+        Toroidal radius at which to constrain the plasma midplane.
+    scale: float
+        Current scale with which to calculate the constraints
+    inboard: bool (default=True)
+        Boolean controlling whether to constrain the inboard (if True) or
+        outboard (if False) side of the plasma midplane.
+    """
+    eq.coilset.set_control_currents(vector * scale)
+    lcfs = eq.get_LCFS()
+    if inboard:
+        constraint[:] = radius - min(lcfs.x)
+    else:
+        constraint[:] = max(lcfs.x) - radius
+    return constraint
+
+
+def coil_force_constraints(
```

Comment:
This looks like three constraints combined into one - for PF_Fz, for CS_Fz_sum, and for CS_FZ_sep_max.
Any reason not to split this up?

Reply:
Speed, mostly! I did think about separating these out into individual constraint functions, but then we're left with updated the force response matrices 3 times instead of once.

Reply:
This is fair, but isn't user-facing or urgent. I will open a separate issue and try and address it in this feature branch, but if not down the line

Reply:
see #1107